src/storage/src/streaming_table/state_table.rs

```diff
@@ -0,0 +1,451 @@
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::ops::Index;
+use std::sync::Arc;
+
+use bytes::BufMut;
+use itertools::Itertools;
+use risingwave_common::array::Row;
+use risingwave_common::buffer::Bitmap;
+use risingwave_common::catalog::{ColumnDesc, TableId, TableOption};
+use risingwave_common::error::RwError;
+use risingwave_common::types::VirtualNode;
+use risingwave_common::util::hash_util::CRC32FastBuilder;
+use risingwave_common::util::ordered::OrderedRowSerializer;
+use risingwave_common::util::sort_util::OrderType;
+use risingwave_pb::catalog::Table;
+
+use super::mem_table::{MemTable, RowOp};
+use super::Distribution;
+use crate::error::{StorageError, StorageResult};
+use crate::row_serde::{
+    serialize_pk, ColumnDescMapping, RowBasedSerde, RowDeserialize, RowSerde, RowSerialize,
+};
+use crate::storage_value::StorageValue;
+use crate::store::{ReadOptions, WriteOptions};
+use crate::{Keyspace, StateStore};
+
+/// `StateTable` is the interface accessing relational data in KV(`StateStore`) with
+/// row-based encoding.
+pub type StateTable<S> = StateTableBase<S, RowBasedSerde>;
+/// `StateTableBase` is the interface accessing relational data in KV(`StateStore`) with
+/// encoding, using `RowSerde` for row to KV entries.
+///
+/// For tables without distribution (singleton), the `DEFAULT_VNODE` is encoded.
+pub const DEFAULT_VNODE: VirtualNode = 0;
+#[derive(Clone)]
+pub struct StateTableBase<S: StateStore, RS: RowSerde> {
+    /// buffer row operations.
+    mem_table: MemTable,
+
+    /// write into state store.
+    keyspace: Keyspace<S>,
+
+    /// All columns of this table. Note that this is different from the output columns in
+    /// `mapping.output_columns`.
+    table_columns: Vec<ColumnDesc>,
+
+    /// Used for serializing the primary key.
+    pk_serializer: OrderedRowSerializer,
+
+    /// Used for serializing the row.
+    row_serializer: RS::Serializer,
+
+    /// Mapping from column id to column index. Used for deserializing the row.
+    mapping: Arc<ColumnDescMapping>,
+
+    /// Indices of primary key.
+    /// Note that the index is based on the all columns of the table, instead of the output ones.
+    // FIXME: revisit constructions and usages.
+    pk_indices: Vec<usize>,
+
+    /// Indices of distribution key for computing vnode.
+    /// Note that the index is based on the all columns of the table, instead of the output ones.
+    // FIXME: revisit constructions and usages.
+    dist_key_indices: Vec<usize>,
+
+    /// Indices of distribution key for computing vnode.
+    /// Note that the index is based on the primary key columns by `pk_indices`.
+    dist_key_in_pk_indices: Vec<usize>,
+
+    /// Virtual nodes that the table is partitioned into.
+    ///
+    /// Only the rows whose vnode of the primary key is in this set will be visible to the
+    /// executor. For READ_WRITE instances, the table will also check whether the writed rows
```

Comment:
Now StateTable is always READ_WRITE, please modify the comment accordingly.

Reply:
Will update these comments later.